### PR TITLE
Fix flaky `test_delayed_replica_failover`

### LINKED ISSUE #54502: Flaky test_delayed_replica_failover
**Describe the bug**
https://s3.amazonaws.com/clickhouse-test-reports/0/9bbd81ca6c27c3645be0d12ee61f304d1494bbe3/integration_tests__release__[4_4].html

**Error message and/or stacktrace**
```
            # If we forbid stale replicas, the query must fail. But sometimes we must have bigger timeouts.
            for _ in range(20):
                try:
                    instance_with_dist_table.query(
                        """
    SELECT count() FROM distributed SETTINGS
        load_balancing='in_order',
        max_replica_delay_for_distributed_queries=1,
        fallback_to_stale_replicas_for_distributed_queries=0
    """
                    )
                    time.sleep(0.5)
                except:
                    break
            else:
>               raise Exception("Didn't raise when stale replicas are not allowed")
E               Exception: Didn't raise when stale replicas are not allowed

test_delayed_replica_failover/test.py:151: Exception
```

### PR DESCRIPTION
**Pull request by serxa**
We now have retries with backoff in ZK client after [commit](https://github.com/ClickHouse/ClickHouse/commit/5947e5b5a85ac29d7cb6f7c2a377cda016aa8b64). And because the test uses `system.zookeeper` to check that Keeper is not available anymore, the test always runs for ~10 minutes and sometimes hits integration test timeout.

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

This closes #54502.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Review by evillique — APPROVED** [review]

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit a45027f22c73b321f1f9cc110f0782fac10d5748 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67541/a45027f22c73b321f1f9cc110f0782fac10d5748/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/363988206511c75777f77178cde566e027db373b/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/601c9d9694af8542d977636cb3bc8cc8d141d570/fast_test.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/601c9d9694af8542d977636cb3bc8cc8d141d570/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by alexey-milovidov** [comment]
@serxa it failed one out of five times.

**Comment by serxa** [comment]
Okay, now it is 12.67 sec instead of 600s. So it is a different problem now. I'll also try to fix this one.

**Comment by serxa** [comment]
> Integration tests flaky check (asan) — Failed tests found: error: 0, passed: 4, fail: 1

We see that 4 runs are successful and the last one, 5th run, is not successful.

The way a flaky check works: it does not restart the server or wipe all the data. Instead it just runs the same test 5 times on the same server and data. So it looks like not all the tests are ready for such an approach.

TLDR. root cause: a part from the previous run was fetched (22:32:01) before we disconnected node_2_1 and node_2_2 and inserted new data (22:32:02) to be fetched

Test logs:

```python
# Hinder replication between replicas of the same shard, but leave the possibility of distributed connection.
pm.partition_instances(node_1_1, node_1_2, port=9009)
pm.partition_instances(node_2_1, node_2_2, port=9009)

node_1_2.query("INSERT INTO replicated VALUES ('2017-05-08', 1)")
node_2_2.query("INSERT INTO replicated VALUES ('2017-05-08', 2)")

time.sleep(1)  # accrue replica delay

assert node_1_1.query("SELECT sum(x) FROM replicated").strip() == "0"
assert node_1_2.query("SELECT sum(x) FROM replicated").strip() == "1"
assert node_2_1.query("SELECT sum(x) FROM replicated").strip() == "0"
assert node_2_2.query("SELECT sum(x) FROM replicated").strip() == "2"
```

logs for iptables:
```
2024-08-04 22:32:01 [ 568 ] DEBUG : [network] 25eb7e1a6e13aa52d27b26ca0b5ca4db844787a386fc45ff5bddad4cf7d1f487: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.1.5', '-d', '172.16.1.7', '--dport', '9009', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
2024-08-04 22:32:01 [ 568 ] DEBUG : [network] 25eb7e1a6e13aa52d27b26ca0b5ca4db844787a386fc45ff5bddad4cf7d1f487: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.1.7', '-d', '172.16.1.5', '--dport', '9009', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
2024-08-04 22:32:01 [ 568 ] DEBUG : [network] 25eb7e1a6e13aa52d27b26ca0b5ca4db844787a386fc45ff5bddad4cf7d1f487: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.1.8', '-d', '172.16.1.6', '--dport', '9009', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
2024-08-04 22:32:01 [ 568 ] DEBUG : [network] 25eb7e1a6e13aa52d27b26ca0b5ca4db844787a386fc45ff5bddad4cf7d1f487: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.1.6', '-d', '172.16.1.8', '--dport', '9009', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
```

data insertion:
```
2024-08-04 22:32:02 [ 568 ] DEBUG : Executing query INSERT INTO replicated VALUES ('2017-05-08', 2) on node_2_2 (cluster.py:3517, query)
```

node_2_1 logs for fetching part that should not be fetched because connection between node_2_1 <-> node_2_2 should be broken (and data was not even inserted yet, so it looks like it reads data from a previous run):
```
2024.08.04 22:32:01.246382 [ 597 ] {} <Information> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Code: 234. DB::Exception: No active replica has part 201705_0_0_0 or covering part (cannot execute queue-0000000000: GET_PART with virtual parts [201705_0_0_0]). (NO_REPLICA_HAS_PART)
2024.08.04 22:32:01.246620 [ 455 ] {} <Trace> default.replicated (ReplicatedMergeTreePartCheckThread): Part 201705_0_0_0 in zookeeper: false, locally: false
2024.08.04 22:32:01.246685 [ 455 ] {} <Information> default.replicated (ReplicatedMergeTreePartCheckThread): Checking if anyone has a part 201705_0_0_0 or covering part.
2024.08.04 22:32:01.247018 [ 455 ] {} <Information> default.replicated (ReplicatedMergeTreePartCheckThread): Found the missing part 201705_0_0_0 at 201705_0_0_0 on node_2_2
2024.08.04 22:32:01.443363 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Fetching part 201705_0_0_0 from default:/clickhouse/tables/2/replicated/replicas/node_2_2
2024.08.04 22:32:01.445225 [ 613 ] {} <Test> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Disk for fetch is not provided, reserving space using storage balanced reservation
2024.08.04 22:32:01.445261 [ 613 ] {} <Test> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Disk for fetch is not provided, reserving space using TTL rules
2024.08.04 22:32:01.445309 [ 613 ] {} <Trace> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Trying to reserve 1.00 MiB using storage policy from min volume index 0
2024.08.04 22:32:01.445379 [ 613 ] {} <Trace> DiskLocal: Reserved 1.00 MiB on local disk `default`, having unreserved 203.51 GiB.
2024.08.04 22:32:01.445410 [ 613 ] {} <Trace> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Disk for fetch is not provided, getting disk from reservation default with type 'local'
2024.08.04 22:32:01.445461 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Downloading part 201705_0_0_0 onto disk default.
2024.08.04 22:32:01.445554 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Downloading files 11
2024.08.04 22:32:01.448969 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (Fetcher): Download of part 201705_0_0_0 onto disk default finished.
2024.08.04 22:32:01.449130 [ 613 ] {} <Trace> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Renaming temporary part tmp-fetch_201705_0_0_0 to 201705_0_0_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
2024.08.04 22:32:01.449237 [ 613 ] {} <Test> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): preparePartForCommit: inserting 201705_0_0_0 (state PreActive) into data_parts_indexes
2024.08.04 22:32:01.449268 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Committing part 201705_0_0_0 to zookeeper
2024.08.04 22:32:01.449850 [ 665 ] {} <Trace> virtual void Coordination::ZooKeeperCreateRequest::writeImpl(WriteBuffer &) const: Creating part at path /clickhouse/tables/2/replicated/replicas/node_2_1/parts/201705_0_0_0
2024.08.04 22:32:01.451558 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Part 201705_0_0_0 committed to zookeeper
2024.08.04 22:32:01.451664 [ 613 ] {} <Debug> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Fetched part 201705_0_0_0 from default:/clickhouse/tables/2/replicated/replicas/node_2_2
2024.08.04 22:32:01.451737 [ 485 ] {} <Test> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (MergerMutator): Nothing to merge in partition 201705: Did not find any parts to merge (with usual merge selectors)
2024.08.04 22:32:01.451749 [ 613 ] {} <Test> default.replicated (ReplicatedMergeTreeQueue): Removing successful entry queue-0000000000 from queue with type GET_PART with virtual parts [201705_0_0_0]
2024.08.04 22:32:01.451796 [ 613 ] {} <Test> default.replicated (ReplicatedMergeTreeQueue): Adding parts [201705_0_0_0] to current parts
2024.08.04 22:32:01.451800 [ 485 ] {} <Trace> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6) (MergerMutator): Checked 1 partitions, found 0 partitions with parts that may be merged: [] (max_total_size_to_merge=109256116224, merge_with_ttl_allowed=true)
2024.08.04 22:32:01.451835 [ 613 ] {} <Test> default.replicated (ReplicatedMergeTreeQueue): Removing part 201705_0_0_0 from mutations (remove_part: false, remove_covered_parts: true)
2024.08.04 22:32:01.451873 [ 485 ] {} <Trace> default.replicated (ac6c2fad-3433-43d7-bf45-acafb9c9b7a6): Scheduling next merge selecting task after 34568ms, current attempt status: CannotSelect
```
